cranelift/zkasm_data/benchmarks/Design.md

```diff
@@ -0,0 +1,142 @@
+
+- **Analysis** targets understanding the performance of zkASM generated by the compiler and tries to answer questions like: Which instructions are executed how often during execution and which are hot instructions? How is performance affected by changes to the compiler? How can the performance of generated zkASM be improved?
+- **Comparison** relates the performance of ZK WASM to the performance of other ZK proving systems and tries to answer questions like: Which proving system is more performant? What are the reasons for differences in performance?
+
+# Analysis
+
+This section describes the technical architecture and user interface of infrastructure and tooling related to *analysis*.
+
+## Instrumentation
+
+Instrumentation is based on logging zkASM instructions by calling Javascript from zkASM. The indirection via Javascript is required as `zkevm-proverjs` executes compiled PIL instead of zkASM. The code for zkASM instruction logging might be inserted to zkASM by enabling a compilation flag or in post processing.
+
+Subsequently raw instruction logs are processed to aggregate data and surface important information. Information to be aggregated includes the number of times an instruction was executed and the number of register allocations. Analyzing the number of register allocations is relevant as [this example](https://github.com/near/wasmtime/issues/181) shows how reducing register allocations can decrease the number of cycles.
+
+TODO specify data to be logged after we committed to the visualization scheme.
+
+## Visualization
+
+Visualizations should help to make the data described above more easily comprehensible and highlight the most expensive opcodes. The following graph shows the schema of the visualizations to be produced for *Analysis*:
+
+![Analysis graph](/cranelift/zkasm_data/benchmarks/assets/analysis_graph.svg)
+
+The graph represents an ordered list of the program's [`MInst`], which can be found on the bottom. Above each `MInst` are the zkASM instructions that have been executed due to its lowering. The following characteristics should hold for the visualization:
+
+- Ordered by decreasing `<count>` from left to right.
+- `<count>` is defined as a tuple `(a, b, c)` of the elements `{num_occurrences, num_cycles, num_reg_writes}`.
+    - `num_occurrences` is the number of times the instruction was executed.
+    - `num_cycles` is the total number of virtual machine cycles required to execute the instruction.
+    - `num_reg_writes` is the total number of register writes caused by the execution of the instruction.
+    - For `zkasm_op_x` sums are taken separately for each `MInst` that emitted `zkasm_op_x`.
+- The order of the tuple elements can be chosen via an UI or CLI flags. For example, `num_cycles` can be shown first, in which case `a = num_cycles`.
+- Sorting is done by comparing the first element of `<count>` tuples.
+- All `zkasm_op_*` rectangles have the same width.
```

Comment:
What was your motivation for this two-level breakdown?

The more I think about this, the more it seems that the further breakdown of `MInst::Instruction` into `zkasm_op_*` will not bring much value compared to the complexity it will introduce.
The lowering from the concrete `MInst` to ZKASM opcodes is fairly deterministic and can be seen in the ZKASM backend code. I think we'll all soon get the intuition about how much each instruction costs and would not need to dive into this detail when doing an optimization of the benchmark like SHA256.
The only scenario where we want to see this breakdown is when we optimize a specific instruction but in this case, it's enough for us to see a breakdown only for this instruction and not for all instructions in the program.

It might be useful to think end-to-end about the common optimization journeys and what information we need to be able to do them. Two that come to mind right now are:
- Make a single benchmark faster
- Make a particular instruction faster

Reply:
## How to use flame graphs?

I assume flame graphs are typically used to see where a program spends most time and then try to optimize that work. For instance, a web frontend might be slow because it spends a lot of time drawing colors on the screen. Further examinations show mostly blue rectangles are re-colored blue, which is unnecessary. So the frontend is optimized by drawing a color only if the area currently has another color.

The compiler receives wasm as input and generates zkASM. We also want to see where most of the time is spent and then optimize that work. However, our understanding of the work done by the program is not as deep as in the previous example and therefore we cannot change what the program does (draw a color only if it differs from the area’s current color). At least in the general case, i.e. when the program is not a benchmark for which we control the original source code.

The wasm we receive as input likely already has been optimized by the compiler that produced it, since developers are not expected to write larger programs directly in wasm. This makes it even harder to figure out what the program does.

For these reasons I’m wondering if (conventional) flame graphs contain a lot of information that we cannot utilize to optimize the emitted zkASM? In that case something simpler might already be sufficient. Or perhaps I am missing something or having a wrong understanding of something?

## Usage of *Analysis* benchmarking tools

My assumption was that *Analysis* benchmarking tools will be used to identify:

1. Hot `MInst`s for a given program and a breakdown of their opcodes and costs on the zkASM side. This helps to identify what to optimize since even micro optimizations of hot opcodes can have a significant impact overall.
2. Opcode sequences that lend themselves to peephole optimizations. For this task I think it would be helpful to see which blocks/labels of a program are on the hot path, since optimizations here are likely to have a higher impact.

## Motivation for graph schema in 8c2bcf0

The motivation was to help with 1. described above. Looking at WAT files like the [sha256 benchmark](https://github.com/near/wasmtime/blob/main/cranelift/zkasm_data/benchmarks/sha256/from_rust.wat) doesn’t reveal which `MInst`s or zkASM opcodes are hot. The graph aggregates `<count>` values over the execution of the program and hence helps to identify hot `MInst`s and zkASM opcodes.

Having the breakdown of `MInst` to zkASM opcodes in the graph helps to make information available in one place. Though as you mentioned the zkASM opcodes can be looked up in backend code and including them in the graph might not be worth the extra complexity.

## Another visualization: hot labels

The visualization proposed above wouldn’t help with 2. An idea for a visualization that helps with identifying peephole optimization opportunities would be printing hot labels. Below is a rough sketch of how that might look. There could be CLI flags to determine how much data is included, e.g. to include only the X most frequent instructions per block to avoid cluttering the graph.

![labels_graph](https://github.com/near/wasmtime/assets/32731489/dd45a6b0-d223-4465-a21e-580e0ff5a5fc)

This shows hot labels for a given program and would allow us to identify the labels where peephole optimizations might have the biggest impact on performance. It could also help with the “make a single benchmark faster” optimization journey.

## Flame graphs via wasm profiling

If control flow of wasm and the zkASM it gets compiled to are comparable, we might consider producing flame graphs by profiling wasm execution. It is [supported by wasmtime](https://docs.wasmtime.dev/examples-profiling-perf.html) and there are many libraries to generate flame graphs from standard profiling data (e.g. [flamegraph](https://github.com/flamegraph-rs/flamegraph) for Rust). This procedure could be used for the benchmark optimization journey where we know and control the Rust and wasm of the program that is compiled to zkASM.

Reply:
> The compiler receives wasm as input and generates zkASM. We also want to see where most of the time is spent and then optimize that work. However, our understanding of the work done by the program is not as deep as in the previous example and therefore we cannot change what the program does (draw a color only if it differs from the area’s current color). At least in the general case, i.e. when the program is not a benchmark for which we control the original source code.

Thanks for raising this very important point. When we build the tooling for performance optimization, we need to have a clear set of use cases we aim to address. For Stage 2, those use cases are:

1. Support and optimize benchmarks **equivalent** to https://risc0.github.io/ghpages/dev/benchmarks/index.html for the purpose of fair comparison
2. Support and optimize [full WASM interpreter](https://github.com/near/wasmtime/issues/174)

In both cases, we do have control over the code that will be compiled to WASM and later to ZKASM and I would expect us to change that code to yield better performance (like we already do for benchmarks by compiling them with `no_std`). We also can annotate this code with tracing statements (e.g. using [`tracing`](https://docs.rs/tracing/latest/tracing/)) to make it easier to see the correspondence between the original and generated code.

In other words, we're not aiming to build tooling to optimize arbitrary ZKASM programs in the wild, instead, we focus on a selected subset of "pet" programs that are important to us.
Note, that this will also be the case beyond Stage 2 - in Stage 3 we will do optimization specifically for NEAR Protocol code and NEAR host functions. Similarly, any users that will be using ZKWASM will have a particular Rust program that they are trying to produce ZK proofs for (e.g. a NEAR light client) and they will be willing to optimize this Rust program and will also benefit from such tooling.

In this light, I think we should focus on visualizations, that allow us to work effectively with programs that have medium-to-large sizes and are annotated/structured well enough to allow us to efficiently optimize them.